akd_audit_quorum/src/crypto/mod.rs

```diff
@@ -0,0 +1,309 @@
+        node_id: NodeId,
+    ) -> Result<EncryptedQuorumKeyShard, QuorumOperationError>;
+
+    /// Save this node's shard of the quorum key to persistent secure storage
+    async fn update_qk_shard(
+        &self,
+        shard: EncryptedQuorumKeyShard,
+    ) -> Result<(), QuorumOperationError>;
+
+    /// Generate the encrypted shards for nodes [0, node_public_keys.len()) encrypted with the public keys of the nodes.
+    /// The provided shards are the quorum key shards, encrypted with THIS node's public key as the leader.
+    async fn generate_encrypted_shards(
+        &self,
+        shards: Vec<EncryptedQuorumKeyShard>,
+        node_public_keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<EncryptedQuorumKeyShard>, QuorumOperationError>;
+
+    /// Encrypt the given material using the provided public key, optionally with the provided nonce
+    async fn encrypt_message(
+        &self,
+        public_key: Vec<u8>,
+        plaintext: Vec<u8>,
+        nonce: Nonce,
+    ) -> Result<Vec<u8>, QuorumOperationError>;
+
+    /// Decrypt the specified message utilizing this node's
+    /// private key. This will require that the message contain
+    /// a nonce. I.e. this cannot be used to decrypt the encrypted
+    /// shard partials
+    async fn decrypt_message(
+        &self,
+        cyphertext: Vec<u8>,
+    ) -> Result<(Vec<u8>, Nonce), QuorumOperationError>;
+
+    /// Generate a commitment on the epoch changes using the quorum key
+    async fn generate_commitment<H: Hasher>(
+        &self,
+        quorum_key_shards: Vec<EncryptedQuorumKeyShard>,
+        epoch: u64,
+        previous_hash: H::Digest,
+        current_hash: H::Digest,
+    ) -> Result<QuorumCommitment<H>, QuorumOperationError>;
+
+    /// Validate the commitment applied on the specified epoch settings
+    async fn validate_commitment<H: Hasher>(
+        public_key: Vec<u8>,
+        commitment: QuorumCommitment<H>,
+    ) -> Result<bool, QuorumOperationError>;
+
+    // ==================================================================
+    // Common trait logic
+    // ==================================================================
+
+    /// Get the number of shards required to reconstruct the quorum key
+    fn shards_required(n: u8) -> u8 {
+        let f = if (n - 1) % 3 == 0 {
+            (n - 1) / 3
+        } else {
+            // there's remainders
+            1 + (n - 1) / 3
+        };
+        2 * f + 1
+    }
+
+    /// Generate num_shards shards of the quorum key, and return the shard pieces.
+    /// We take ownership of the quorum key here to help prevent leakage of the key.
+    /// By taking ownership, someone needs to explicitely clone it to use it elsewhere
+    fn generate_shards(
+        quorum_key: [u8; QUORUM_KEY_SIZE],
+        num_shards: u8,
+    ) -> Result<Vec<QuorumKeyShard>, QuorumOperationError> {
+        let num_approvals = Self::shards_required(num_shards);
+
+        let mut parts = vec![vec![]; num_shards.into()];
```

Comment:
Is `shards` a better name instead of `parts`? A part below is split into num_shard shards and this array stores all i-th shards in its [i] element.